LOPS-741 - PantheonSecrets Documentation (#8975)
pantheon-systems/documentation: source/content/guides/secrets/04-use-cases.md

---
title: Pantheon Secrets Guide
subtitle: Use Cases
description: Some common uses cases for Pantheon Secrets
terminuspage: true
contributors: [stovak]
contenttype: [guide]
innav: [true]
categories: [secrets]
cms: [drupal, wordpress]
audience: [development]
product: [secrets]
integration: [--]
tags: [reference, cli, local, terminus, workflow]
permalink: docs/guides/secrets/use-cases
reviewed: "2024-05-01"
---

> # Use Case: Using secrets with Integrated Composer

Reviewer: All gone!

## Mechanism 1: Oauth composer authentication (recommended)

#### GitHub Repository

1. [Generate a Github token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token). The Github token must have all "repo" permissions selected.

    NOTE: Check the repo box that selects all child boxes. **Do not** check all child boxes individually as this does not set the correct permissions.

    ![image](https://user-images.githubusercontent.com/87093053/191616923-67732035-08aa-41c3-9a69-4d954ca02560.png) 

> 1. Set the secret value to the token via terminus: `terminus secret:site:set <site> github-oauth.github.com <github_token> --type=composer --scope=user,ic`

Reviewer 2: It would be better if the Terminus command was a block comment, like the other examples, instead of being an inline comment. Otherwise, it is hard to find when scanning the page.

We probably shouldn't include the "user" scope in examples.

Author: https://github.com/pantheon-systems/documentation/pull/8975/commits/134dce894236acac146d001ae55318dbbabef568

1. Add your private repository to the `repositories` section of `composer.json`:

    ```json
    {
        "type": "vcs",
        "url": "https://github.com/your-organization/your-repository-name"
    }
    ```

>     Your repository should contain a `composer.json` that declares a package name in its `name` field. If it is a WordPress plugin or a Drupal module, it should specify a `type` of `wordpress-plugin` or `drupal-module` respectively. For these instructions, we will assume your package name is `your-organization/your-package-name`.

Reviewer: > Your repository should contain a `composer.json` that declares a package name in its `name` field. It should specify a `type` like  `wordpress-plugin` or `drupal-module` for example. For these instructions, we will assume your package name is `your-organization/your-package-name`.

1. Require the package defined by your private repository's `composer.json` by either adding a new record to the `require` section of the site's `composer.json` or with a `composer require` command:

    ```bash
    composer require your-organization/your-package-name
    ```

1. Commit your changes and push to Pantheon.

> #### GitLab Repository

Reviewer 2: Maybe Gitlab and Bitbucket examples can become tabs, along with the previous "Github" section. Otherwise, the content below is too repetitive.

Author: I tried to get them to work, but they don't seem to be working right. I'll get Rachel to correct my problem, but the tabs are there delineated.

Author: https://github.com/pantheon-systems/documentation/pull/8975/commits/940130004448dcdf995cad6633822238b93986a9

1. Set the secret value to the token via Terminus: `terminus secret:site:set <site> gitlab-oauth.gitlab.com <gitlab_token> --type=composer --scope=user,ic`

1. Add your private repository to the `repositories` section of `composer.json`:

    ```json
    {
        "type": "vcs",
        "url": "https://gitlab.com/your-group/your-repository-name"
    }
    ```

    Your repository should contain a `composer.json` that declares a package name in its `name` field. If it is a WordPress plugin or a Drupal module, it should specify a `type` of `wordpress-plugin` or `drupal-module` respectively. For these instructions, we will assume your package name is `your-organization/your-package-name`.

1. Require the package defined by your private repository's `composer.json` by either adding a new record to the `require` section of the site's `composer.json` or with a `composer require` command:

    ```bash
    composer require your-group/your-package-name
    ```

1. Commit your changes and push to Pantheon.

#### Bitbucket Repository

> 1. [Generate a Bitbucket oauth consumer](https://support.atlassian.com/bitbucket-cloud/docs/use-oauth-on-bitbucket-cloud/). Ensure that Read repositories permission is selected for the consumer. Also, set the consumer as private and put a (dummy) callback URL.

Reviewer: Done.

`terminus secret:site:set ${SITE_NAME} COMPOSER_AUTH ${COMPOSER_AUTH_JSON} --type=env --scope=user,ic`
```

# Use Case: Using secrets with Drupal Key module
If you want to use Pantheon Secrets in your Drupal application through the [Key module](https://www.drupal.org/project/key), you should use the [Pantheon Secrets](https://www.drupal.org/project/pantheon_secrets) module.

# Pantheon Secrets detailed example

Please look at the [module documentation](../README.md) if you have not done so yet.

In this guide we will go over an end to end example on how to setup secrets for a given site and how to use those secrets on a module that integrates with the Key module. For this example, we will use the [Sendgrid API](https://www.drupal.org/project/sendgrid_api) and [Sendgrid Mailer](https://www.drupal.org/project/sendgrid_mailer) modules.

## Prerequisites

1) Make sure you have access to a Drupal >= 9.4 site running PHP >= 8.0 hosted on Pantheon.

1) Make sure you have [terminus installed](https://docs.pantheon.io/terminus/install#install-terminus) in your machine

1) Install the [Secrets Manager Plugin](https://github.com/pantheon-systems/terminus-secrets-manager-plugin#installation)

> 1) Install the required modules in your Drupal site and push the changes to Pantheon:

Reviewer: Ordered lists fixed.

    ```
    composer require drupal/pantheon_secrets drupal/sendgrid_api drupal/sendgrid_mailer
    git add composer.json composer.lock
    git commit -m "Add required modules."
    git push
    ```

1) Enable the modules:
    ```
    terminus drush <site>.<env> -- en -y pantheon_secrets sendgrid_api sendgrid_mailer
    ```

1) Make sure your Sendgrid account is correctly configured and allows sending email.

1) Create a Sendgrid API key by following [Sendgrid instructions](https://docs.sendgrid.com/ui/account-and-settings/api-keys#creating-an-api-key)

> 1) Store the API key as a site secret:

Reviewer: Lists fixed.

>     ```

Reviewer: I think I got it fixed.

    terminus secret:site:set <site> sendgrid_api <api_key> --scope=web --type=runtime
    ```

>     You can optionally add a [secret environment override](https://github.com/pantheon-systems/terminus-secrets-manager-plugin#environment-override) to change the API key for a given environment (e.g. you want to use different Sendgrid accounts for live and dev environments)

Reviewer: It now reads:
> As a best practice, the non-production environments should be the default and then override that value with a [secret environment override](https://github.com/pantheon-systems/terminus-secrets-manager-plugin#environment-override) to change the API key for the live environment (e.g. you want to use different Sendgrid accounts for live and dev environments)

>     You can optionally add a [secret environment override](https://github.com/pantheon-systems/terminus-secrets-manager-plugin#environment-override) to change the API key for a given environment (e.g. you want to use different Sendgrid accounts for live and dev environments)

Reviewer: Oh, and I think the tab isn't an issue anymore?

1) Add the Key entity in one of the different available ways:

    1) Add a new key through the Key module UI. Select Pantheon Secret as the key provider and your secret name from the dropdown (make sure you select "Sendgrid" as the Key type and "Pantheon" as the Key provider)

        ![Screenshot of creating a new Key entity with type "Sendgrid" and provider "Pantheon"](add-key.png)

>     1) Go to /admin/config/system/keys/pantheon and click on the "Sync Keys" button to get all of the available secrets into Key entities.

Reviewer 3: use code formatting for paths

```suggestion
    1) Go to `/admin/config/system/keys/pantheon` and click on the "Sync Keys" button to get all of the available secrets into Key entities.
```

Reviewer: Added.

        ![Screenshot of Sync Pantheon Secrets page in Drupal UI](../../../images/sync-keys.png)

>         Then, edit the sendgrid_api Key and change the type to "Sendgrid"

Reviewer 3: ```suggestion
        Then, edit the `sendgrid_api` Key and change the type to "Sendgrid"
```

    1) Use the provided drush command to sync all of your secrets into Key entities:
        ```
        terminus drush <site>.<env> -- pantheon-secrets:sync
        ```

>         Then, edit the sendgrid_api Key and change the type to "Sendgrid"

Reviewer 3: ```suggestion
        Then, edit the `sendgrid_api` Key and change the type to "Sendgrid"
```

> 1) Go to the Sendgrid API Configuration page (/admin/config/services/sendgrid) and select your Key item

Reviewer 3: ```suggestion
1) Go to the Sendgrid API Configuration page (`/admin/config/services/sendgrid`) and select your Key item
```

Reviewer: Added.

    ![Screenshot of Sendgrid API Configuration page in Drupal UI](../../../images/sendgrid-config.png)

> 1) Make sure your site "Email Address" (/admin/config/system/site-information) matches a verified Sender Identity in Sendgrid

Reviewer 3: ```suggestion
1) Make sure your site "Email Address" (`/admin/config/system/site-information`) matches a verified Sender Identity in Sendgrid
```

> 1) Go to the Sendgrid email test page (/admin/config/services/sendgrid/test) and test your Sendgrid integration by sending a test email

Reviewer 3: ```suggestion
1) Go to the Sendgrid email test page (`/admin/config/services/sendgrid/test`) and test your Sendgrid integration by sending a test email
```

    ![Screenshot of Sendgrid email test page in Drupal UI](../../../images/sendgrid-email-test.png)

> 1) The email should get to your inbox. Enjoy!

Reviewer: Fixed up the ordered list formatting and added backticks.

# Use Case: Accessing secrets from your codebase

### Introduction

> Include this note: "Note: Only get has been implemented so far. You should handle your secrets through terminus using [Terminus Secrets Manager](https://github.com/pantheon-systems/terminus-secrets-manager-plugin)." Do not present this as something we "may" do in the future!

Reviewer 2: This section needs to be written. Follow the advice of the note. Only document "get", say setting secrets from the application is not supported.

Reviewer: I reworded this somewhat.  It still needs work.

> Note: Only GET has been implemented. Secrets are created and modified via the [Terminus Secrets Manager Plugin](https://github.com/pantheon-systems/terminus-secrets-manager-plugin). Modifying secrets from within the application itself is not supported.
> 
> This SDK will only read secrets with scope web. Secrets get cached in the server for 15 minutes so you should wait (at most) that time if you modified your site secrets.
> 
> Note: this also applies to quicksilver scripts

Reviewer: We made more changes while discussing on Zoom.  Marking resolved.

> Also: [https://github.com/pantheon-systems/customer-secrets-php-sdk?tab=readme-ov-file#restrictions](https://github.com/pantheon-systems/customer-secrets-php-sdk?tab=readme-ov-file#restrictions)

Reviewer 3: This line is missing context. Was it meant to be included verbatim?

Reviewer: Removed. :)

1. Set the secret value to the consumer info via Terminus: 
   ```bash
   terminus secret:site:set <site> bitbucket-oauth.bitbucket.org "<consumer_key> <consumer_secret>" --type=composer --scope=user,ic`
   ```

1. Add your private repository to the `repositories` section of `composer.json`:

    ```json
    {
        "type": "vcs",
        "url": "https://bitbucket.org/your-organization/your-repository-name"
    }
    ```

    Your repository should contain a `composer.json` that declares a package name in its `name` field. If it is a WordPress plugin or a Drupal module, it should specify a `type` of `wordpress-plugin` or `drupal-module` respectively. For these instructions, we will assume your package name is `your-organization/your-package-name`.

1. Require the package defined by your private repository's `composer.json` by either adding a new record to the `require` section of the site's `composer.json` or with a `composer require` command:

    ```bash
    composer require your-organization/your-package-name

>     ```

Reviewer: Removed that indentation.

    ![Screenshot of Sendgrid API Configuration page in Drupal UI](../../../images/sendgrid-config.png)

1) Make sure your site "Email Address" (`/admin/config/system/site-information`) matches a verified Sender Identity in Sendgrid

1) Go to the Sendgrid email test page (`/admin/config/services/sendgrid/test`) and test your Sendgrid integration by sending a test email

    ![Screenshot of Sendgrid email test page in Drupal UI](../../../images/sendgrid-email-test.png)

1) The email should get to your inbox. Enjoy!

# ☞ Accessing secrets from your codebase

### Introduction

Note: Only GET has been implemented so fDK. You should handle your secrets through terminus using [Terminus Secrets Manager](https://github.com/pantheon-systems/terminus-secrets-manager-plugin).

Also: [https://github.com/pantheon-systems/customer-secrets-php-sdk?tab=readme-ov-file#restrictions](https://github.com/pantheon-systems/customer-secrets-php-sdk?tab=readme-ov-file#restrictions)

Note: this also applies to quicksilver scripts

## Mechanism 1: get_pantheon_secrets

## Mechanism 2: OOP (get a better name here!!!)

> [https://github.com/pantheon-systems/customer-secrets-php-sdk?tab=readme-ov-file#usage](https://github.com/pantheon-systems/customer-secrets-php-sdk?tab=readme-ov-file#usage)

Reviewer: Agreed.  I moved it higher up in the document and made it look moderately less incomplete.  It needs a `get_pantheon_secrets()` code example.

https://pr-8975-documentation.appa.pantheon.site/guides/secrets/use-cases#-accessing-secrets-from-your-codebase

</Tab>

</TabList>

## Mechanism 2: HTTP Basic Authentication

You may create a `COMPOSER_AUTH json` and make it available via the `COMPOSER_AUTH` environment variable if you have multiple private repositories on multiple private domains.

Composer has the ability to read private repository access information from the environment variable: `COMPOSER_AUTH`. The `COMPOSER_AUTH` variables must be in a [specific JSON format](https://getcomposer.org/doc/articles/authentication-for-private-packages.md#http-basic).

**Format example:**

```bash
#!/bin/bash

read -e COMPOSER_AUTH_JSON <<< {
    "http-basic": {
        "github.com": {
            "username": "my-username1",
            "password": "my-secret-password1"
        },
        "repo.example2.org": {
            "username": "my-username2",
            "password": "my-secret-password2"
        },
        "private.packagist.org": {
            "username": "my-username2",
            "password": "my-secret-password2"
        }
    }
}
EOF

terminus secret:site:set ${SITE_NAME} COMPOSER_AUTH ${COMPOSER_AUTH_JSON} --type=env --scope=ic
```

# ☞ Accessing secrets from your codebase

## Introduction

> Note: Only GET has been implemented.  Secrets are created and modified via the [Terminus Secrets Manager Plugin](https://github.com/pantheon-systems/terminus-secrets-manager-plugin).  Modifying secrets from within the application itself is not supported.

Reviewer 2: ```suggestion
It is only possible to read secrets from the Drupal or WordPress site. Secrets are created and modified via the [Terminus Secrets Manager Plugin](https://github.com/pantheon-systems/terminus-secrets-manager-plugin).  Modifying secrets from within the application itself is not supported.
```

> This SDK will only read secrets with scope web. Secrets get cached in the server for 15 minutes so you should wait (at most) that time if you modified your site secrets.

Reviewer 2: ```suggestion
Secrets must have the scope `web` to be visible from your application. Secrets are cached in the server for 15 minutes, so you must wait for a while after modifying secret values before they will be available for use.
```

Note: this also applies to quicksilver scripts

## Mechanism 1: SDK (composer package)

This PHP library is the recommended approach for accessing secrets from the Secrets Service.

Use composer to install it:

```
composer require pantheon-systems/customer-secrets-php-sdk
```

### Usage

In your PHP code, do the following:

```
use PantheonSystems\CustomerSecrets\CustomerSecrets;

$client = CustomerSecrets::create()->getClient();
$secret = $client->getSecret('foo');
$secret_value = $secret->getValue();

// You could also get all of your secrets like this:
$secrets = $client->getSecrets();
```

See the included classes and internal documentation for more examples and use cases.
[https://github.com/pantheon-systems/customer-secrets-php-sdk?tab=readme-ov-file#usage](https://github.com/pantheon-systems/customer-secrets-php-sdk?tab=readme-ov-file#usage)

## Mechanism 2: get_pantheon_secrets

> A `pantheon_get_secrets()` function is provided for sites that aren't built with Composer.

Reviewer 2: ```suggestion
For sites that do not use Composer, the function `pantheon_get_secret()` may be used to fetch the value of a single secret.
```

Reviewer: This merge conflicted with a change I made but I changed this language as suggested.
